### DISCUSSION TITLE
Need Help Understanding Line Wrapping Difference Between Browsers and TextMeasurer

### DISCUSSION THREAD

**Discussion by alecrodden** [SixLabors/Fonts#367] [Q&A] [answered]
A specific sentence, in a specific font (Courier Prime Regular), wraps to 3 lines in a given width in modern browsers (Chrome and Firefox tested), but TextMeasurer counts only 2 lines.

**HTML**
```
<!doctype html>
<html>

<head>
    <link rel="preconnect" href="https://fonts.googleapis.com">
    <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
    <link
        href="https://fonts.googleapis.com/css2?family=Courier+Prime:ital,wght@0,400;0,700;1,400;1,700"
        rel="stylesheet">
    <style>
        *,
        *::before,
        *::after {
            box-sizing: border-box;
        }

        * {
            margin: 0;
            padding: 0;
            font-family: "Courier Prime", monospace;
            font-size: 12pt;
            color: black;
            line-height: 1;
        }
    </style>
</head>

<body>
    <div style="width: 3.875in; background-color: lightgray;">
        Leer, but lonesome has fussin&#x27; change a faith. Themself seen and four trample.
    </div>
</body>

</html>
```
***Rendered Output (Firefox)***
![image](https://github.com/SixLabors/Fonts/assets/6413041/118f547a-abbb-40b9-b4fe-61937e1f5575)

**C#**
```
var fontCollection = new FontCollection();
fontCollection.Add("./CourierPrime-Regular.ttf");
var fontFamily = fontCollection.Families.First();
var font = fontFamily.CreateFont(12);

var text = @"Leer, but lonesome has fussin' change a faith. Themself seen and four trample.";

var textOptions = new TextOptions(font)
{
    Dpi = 96, // 1in = 96px
};
var wrappingLengthInInches = 3.875f;
textOptions.WrappingLength = wrappingLengthInInches * textOptions.Dpi;
Console.WriteLine($"Wrapping Length: {wrappingLengthInInches}in ({textOptions.WrappingLength}px)");

var lineCount = TextMeasurer.CountLines(text, textOptions);
Console.WriteLine($"Line Count: {lineCount}");
```
***Console Output***
```
Wrapping Length: 3.875in (372px)
Line Count: 2
```
Am I missing any other options to set? My understanding is that modern browsers and this library should match. The difference ends up being (looking at the layout generated internally by `TextMeasurer`) that `TextMeasurer` doesn't wrap the "a" before " faith", but browsers do, so browsers end up not having enough room for the last word "trample" and so wrap to a third line.

**Comment by JimBobSquarePants** [comment]
I think it simply boils down to accuracy. We measure the glyph points to determine wrapping and if the glyph (before rasterization) fits in those bounds it will not get wrapped. 

Wrapping positions is determined by the algorithm defined by [UAX14](https://www.unicode.org/reports/tr14/) and pass all the provided tests so there are no issues there.

Maybe the browser is less accurate?

![367](https://github.com/SixLabors/Fonts/assets/385879/410e1b4c-5521-4c71-a265-927c60025597)

**Comment by alecrodden** [comment]
Looks like the glyphs in both renderings line up, the library just allows a little bit more room for the "a"?
![image](https://github.com/SixLabors/Fonts/assets/6413041/2eaaa6a6-cc06-4915-b83f-24f266c1ea6a)

Is the difference in accuracy quantifiable? Can I make a small adjustment to get close enough to browser behavior? Unfortunately, I'm in a situation where I need to lay out text server-side (in order to know the exact page count ahead of time), but ultimately the browser renders the text client-side.

**Reply by JimBobSquarePants** [reply]
If you look at your image the browser is treating the width as an exclusive-lower-bounds. The last-colored pixel is at position 371px.

Our output hits 373px following rasterization but that's simply a case of pixel clamping/rounding in the rasterizer itself. 

Unfortunately, I can't comment on the process browsers use, only follow the standards as accurately as I possibly can.  

I'll have a tinker when I have some downtime though to double check my approach, feel free to download the source and have a look yourself.

**Reply by alecrodden** [reply]
Sorry to bother, have you had a chance to look into this?

**Reply by JimBobSquarePants** [reply]
No. Have you?

**Reply by JimBobSquarePants** [reply]
OK. So, I finally got a chance to investigate this. It turns out we were not breaking until too late if the break opportunity was at the very end of a calculated text line. I've opened a PR #380 that fixes this and other issues. 

Six Labors
![367](https://github.com/SixLabors/Fonts/assets/385879/6f8bfe40-4651-4254-a7d0-f467be5473ea)

Firefox
![367-ff](https://github.com/SixLabors/Fonts/assets/385879/254fda6e-ee5d-403a-921a-70c4b2a8ecaa)

**Reply by alecrodden** [reply]
That's awesome, thank you! When will this show up in stable?

**Reply by JimBobSquarePants** [reply]
Likely Monday, I have a couple of low hanging performance optimisations I want to merge first.

**Reply by JimBobSquarePants** [reply, marked as answer]
Today!! 🎉
